tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc

```diff
@@ -0,0 +1,207 @@
+  if (cerr != cudaSuccess) {
+    LOG(FATAL) << "could not get the default CUDA pool : "
+               << cudaGetErrorString(cerr);
+  }
+  VLOG(1) << Name() << " CudaMallocAsync initialized on platform: "
+          << platform_gpu_id.value() << " with pool size of: "
+          << pool_size << " this ptr: " << this;
+  cerr = cudaMemPoolSetAttribute(pool_, cudaMemPoolAttrReleaseThreshold,
+                                 (void*)&pool_size);
+  if (compute_stats) {
+    stats_.bytes_limit = static_cast<int64>(pool_size);
+  } // If not set, it means we do not compute stats.
+  if (cerr != cudaSuccess) {
+    LOG(FATAL) << "could not set the default CUDA pool memory threshold : "
+               << cudaGetErrorString(cerr);
+    pool_ = nullptr;
+  }
+
+  // If in TF_DETERMINISTIC_OPS is set, then make the allocator behave
+  // determistically.
+  bool deterministic_ops = false;
+  TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar("TF_DETERMINISTIC_OPS",
+                                             /*default_val=*/false,
+                                             &deterministic_ops));
+  if (deterministic_ops) {
+    cudaMemPoolSetAttribute(pool_, cudaMemPoolReuseAllowOpportunistic, 0);
+    cudaMemPoolSetAttribute(pool_, cudaMemPoolReuseAllowInternalDependencies, 0);
+  }
+#endif
+
+  VLOG(2) << Name() << " GPUcudaMallocAsyncAllocator PoolSize " << pool_size;
+  if (reserve_memory) {
+    void* ptr = AllocateRaw(0, pool_size);
+    DeallocateRaw(ptr);
+    VLOG(2) << Name() << " GPUcudaMallocAsyncAllocator reserved the pool";
+    ClearStats();
+  }
+}
+
+GPUcudaMallocAsyncAllocator::~GPUcudaMallocAsyncAllocator() {
+  cuStreamDestroy(cuda_stream_);
+}
+
+void* GPUcudaMallocAsyncAllocator::AllocateRaw(size_t alignment,
+                                               size_t num_bytes) {
+#if CUDA_VERSION < 11020 || !defined(GOOGLE_CUDA)
+  return nullptr;
+#else
+  if (pool_ == nullptr) {
+    LOG(FATAL) << "The instantiation of GPUcudaMallocAsyncAllocator failed."
+               << " See previous errors.";
+  }
+  se::cuda::ScopedActivateExecutorContext scoped_activation{stream_exec_};
+  void* rv = 0;
+  cudaError_t res = cudaMallocFromPoolAsync(&rv, num_bytes, pool_, cuda_stream_);
+  if (res != cudaSuccess) {
+    size_t free, total;
+    cudaMemGetInfo(&free, &total);
+    LOG(ERROR) << Name() << " cudaMallocAsync failed to allocate " << num_bytes
+               << " Free Total: " << free << " " << total
+               << ". Error: " << cudaGetErrorString(res)
+               << " \nStats: \n" << stats_.DebugString();
+    return nullptr;
+  }
+
+  // Update stats.
+  if (stats_.bytes_limit.has_value()) {
+    ++stats_.num_allocs;
+    stats_.bytes_in_use += num_bytes;
+    stats_.peak_bytes_in_use =
+        std::max(stats_.peak_bytes_in_use, stats_.bytes_in_use);
+    stats_.largest_alloc_size =
+        std::max<std::size_t>(stats_.largest_alloc_size, num_bytes);
+    size_map_[rv] = num_bytes;
+  }
```

Comment:
Doesn't this block need to be guarded by the lock_?

Same for the RequestedSizes and AllocatedSizes functions?

Reply:
The declaration is:

`absl::flat_hash_map<const void*, size_t> size_map_ GUARDED_BY(lock_);`

I checked at other places in TF and it seems to guard automatically all access to the variable size_map_. Otherwise, there would be other places that would be buggy.

Reply:
Could you provide a link to the other locations where this is being used? I believe that this is a bug and that you should be using the TF_GUARDED_BY macro. This macro has no effect on behavior, instead it adds thread_annotation which should forbid compilation of code that uses the value being guarded without also acquiring the lock.

Reply:
I did a grep an took a random one. 
I just redid one and the one I picked activated the lock manually as you expect.
I'm not sure if I badly understood the code when reading or there is an hidden bug. Sadly I have no hope of finding again the exact place that I looked at last time. There is too many places to check.

In all cases, I updated the code. In 2 places I didn't take it to read a simple value. In those places, I think it was safe to not take it.